Feature/add character tags to message div option

m-petra-fn opened the pull request:
<!-- Put X in the box below to confirm -->

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

---

# What is the reason for a change?

This change adds a toggleable option (user panel, below the theme options) to allow adding character tags as CSS classes for the .mes divs (message divs) of those characters. This way we can use CSS selectors for groups of characters with that tag to easily style their messages all at once. Useful for groups or characters of a similar team or faction.

# What did you do to achieve this?

- Added a new toggle option.
- Added a function that will, if the toggle is active:
  * Get the tags list and tag_map of characters from the user settings.json file.
  * Create a map of tag names by their ids for quicker look up.
  * Process each div that is children of #chat that has the .mes tag to:
    - Get their normalized avatar filename from the avatar src nested on the div.
    - Use that avatar filename to get all tags of that character.
    - Use those ids to get the names of the tags.
    - Normalize the tag names (lowercase them and replace spaces with hyphens)
    - Add all those tag names as classes for the div.
    - Store all previously processed character class names in the chat into a simple cache so that they are quickly retrieved without having to normalize the tags again.
- Added events to reload the page when tag is added or removed from a character and when the option is toggled.

![2025-07-03 03_00_31-SillyTavern](https://github.com/user-attachments/assets/9725bf71-ea5a-499d-8c45-1575bf49568e)
![2025-07-03 03_02_13-2025-07-03 02_58_50-Window png - XnView MP](https://github.com/user-attachments/assets/386ad90b-1e54-4b4a-9339-140e07d5669f)

**IMPORTANT**: There might be other events that I don't know of that might need to reload the chat to reapply tags.

# How would a reviewer test the change?

![image](https://github.com/user-attachments/assets/57c9b550-5010-4ce7-8b38-137cd599b87c)

First, activate the `Add Char Tags to Message Div` toggle option on the User Settings panel. After, create a few characters, some with the same simple tag like Elf. Then, assign this tag to some of the characters and not to others and create a group chat.

Use the custom CSS field or the CSS Snippets Extension to add a class selector like ".elf" and add to it different properties that are easily visible. All characters with the elf tag should be affected, and the chars that don't shouldn't.

Remove and add the tags to see if the chat is automatically refreshed to account for the changes. Try also posting messages and the like to see if the changes are rendered correctly.

Characters with multiple tags can have several CSS cascading, I believe those that are declared last on the CSS field should get precedence.

---

Example CSS affecting message look by tag:

![image](https://github.com/user-attachments/assets/12f43037-4990-464d-a0f4-c6df392e03f6)
![image](https://github.com/user-attachments/assets/a0cbedf2-38a3-493b-9890-afdf0122a09f)

m-petra-fn committed:
Adding toggle.

m-petra-fn committed:
Adding events for the new toggle. Reloading chat when toggled.

m-petra-fn committed:
Adding chat reload when tags are added or removed to char.

m-petra-fn committed:
Adding main applyCharacterTagsToMessageDivs function to add tags to divs.

m-petra-fn committed:
Adding information that the class name is normalized and lowercased to the tooltip.

m-petra-fn committed:
Correcting console.traces to console.debug for minimal console printing.

m-petra-fn committed:
Lint indentation fix.

m-petra-fn committed:
Adding simple cache.

m-petra-fn committed:
Adding reload event when a new message is added to chat.

Cohee1207 commented:
I'd prefer `data-tag` attribute to prevent accidental collisions with system classes.

m-petra-fn commented:
> I'd prefer `data-tag` attribute to prevent accidental collisions with system classes.

Fair enough. I'll make this change tomorrow.

Cohee1207 commented on public/script.js:
```diff
@@ -2830,6 +2837,121 @@ export function scrollChatToBottom() {
+ */
+export function normalizeTagClassName(name) {
+    // Sanitize: return name, lowercased, with spaces replaced by hyphens (vibecoded sanitizing)
+    return name ? name.replace(/\s+/g, '-').toLowerCase() : null;
```
Consider:
https://developer.mozilla.org/en-US/docs/Web/API/CSS/escape_static

With data-tag you wouldn't have to worry about class names

m-petra-fn replied:
You are correct. It was not necessary after using data-tags.

Cohee1207 commented on public/script.js:
```diff
@@ -2830,6 +2837,121 @@ export function scrollChatToBottom() {
+        return null;
+    }
+
+    const avatarFile = avatarSrc.split('file=')[1];
```
Parsing query params with string split is not reliable. Consider:
```js
const url = new URL(avatarSrc, window.location.origin);
const avatarFile = url.searchParams.get('file');
```

m-petra-fn replied:
Added, worked correctly.

m-petra-fn committed:
Changing tags applied from classes to HTML data API.

m-petra-fn committed:
Adding guide showing how to select data attributes with CSS.

m-petra-fn committed:
Fixing trailing comma pointed out by lint.

m-petra-fn commented:
# Changes

- Added suggested changes.
- Tags are now added as `data-char-tags` (includes all tags, their names as is, separated by commas), and `data-char-tag-TAGNAME`.

![image](https://github.com/user-attachments/assets/3ab0923f-6ef9-4a8a-8521-48e055aaa6c1)
![image](https://github.com/user-attachments/assets/e694de8e-aa22-46cb-a51a-d9cef36045f8)

- Created a guide popup when clicking on the info button next to the toggle.

![image](https://github.com/user-attachments/assets/b27a2406-a82e-4003-ba95-47e0793763fd)

OBS: Found an issue where clicking the info button would toggle the checkbox, fixed by using event.preventDefault() on the event. This also happens with the tag list button. Will see if I open a separate PR with the same fix.

Cohee1207 commented on public/scripts/power-user.js:
```diff
@@ -1369,6 +1370,12 @@ function applyTheme(name) {
+        {
+            key: 'add_char_tags_to_message_div',
+            action: () => {
+                $('#add_char_tags_to_message_div').prop('checked', power_user.add_char_tags_to_message_div);
```
Applying a theme won't change the actual state of the message divs

m-petra-fn replied:
I forgot to check what this actually was for when I copied boilerplate of another checkbox.

The toggle was now removed, so this code no longer exists.

Cohee1207 commented on public/scripts/power-user.js:
```diff
@@ -4037,6 +4046,12 @@ $(document).ready(() => {
+    $('#add_char_tags_to_message_div').on('input', function () {
+        power_user.add_char_tags_to_message_div = !!$(this).prop('checked');
+        saveSettingsDebounced();
+        reloadCurrentChat();
```
I'm sure there can be a way around reloading: removing all attributes with a `data-char-tag` prefix.

m-petra-fn replied:
Reload was now improved after your tip. Attributes are destroyed and reconstructed after tag changes.

Cohee1207 commented on public/scripts/tags.js:
```diff
@@ -686,6 +687,10 @@ function selectTag(event, ui, listSelector, { tagListOptions = {} } = {}) {
 
     addTagsToEntity(tag, characterIds, { tagListSelector: listSelector, tagListOptions: tagListOptions });
 
+    // Reload the chat so that tags are applied to the message divs and css can be loaded
```
See the other comment about reloading

m-petra-fn replied:
Reload was now improved after your tip. Attributes are destroyed and reconstructed after tag changes.

Cohee1207 commented on public/scripts/tags.js:
```diff
@@ -1304,6 +1313,17 @@ export function createTagInput(inputSelector, listSelector, tagListOptions = {})
+    html.attr('id', 'tag_data_css_guide');
+    html.append(await renderTemplateAsync('tagDataCSSGuide', {}));
+
+    const tagContainer = $('<div class="tag_data_css_guide ui-sortable"></div>');
```
Why does it have to be here?

m-petra-fn replied:
Originally, I thought this "tag container" was some sort of container for the popup itself, not the tag containers for the tag list that is used on the popup. Because of that I left it as it was.

The toggle was now removed, so this code no longer exists.

Cohee1207 commented on public/scripts/tags.js:
```diff
@@ -2019,6 +2042,14 @@ export function initTags() {
     $(document).on('click', '.tag_view_backup', onTagsBackupClick);
     $(document).on('click', '.tag_view_restore', onBackupRestoreClick);
+
+    $(document).on('click', '#tag_data_css_guide', function (event) {
```
You can replace document with regular id selector.

m-petra-fn replied:
The toggle was now removed, so this code no longer exists.

Cohee1207 commented on public/scripts/templates/tagDataCSSGuide.html:
```diff
@@ -0,0 +1,24 @@
+<div class="title_restorable">
+    <h3 data-i18n="Character Tags CSS Styling Guide">Character Tags CSS Styling Guide</h3>
```
I think it's better be moved to the docs website, and the link pointing to the docs page. For example, in the character creation section.

m-petra-fn replied:
Great idea, I didn't even know the docs were editable. I'll move this there later.

m-petra-fn committed:
Simplifying re-rendering by deleting previous attributes.

m-petra-fn committed:
Adding the feature as default behavior, removing toggles. Will add docs to ST docs.

m-petra-fn committed:
Removing code that is no longer needed.

Cohee1207 commented on public/script.js:
```diff
@@ -2753,6 +2754,8 @@ export function addOneMessage(mes, { type = 'normal', insertAfter = null, scroll
         scrollChatToBottom();
     }
+
+    applyCharacterTagsToMessageDivs();
```
Pass params.mesid to the function and only query for mesid attribute selector if it's present.
That should prevent the exponential effort required to render tags on every consequent message.

m-petra-fn replied:
Added a filter to send a message id or array of ids and only those messages will be re-rendered.

Cohee1207 commented on public/script.js:
```diff
@@ -2830,6 +2833,102 @@ export function scrollChatToBottom() {
+        }
+    });
+
+    console.debug('All data tags removed');
```
Not sure these debugs are needed.

m-petra-fn replied:
Removed and only left a few ones that seemed more appropriate.

Cohee1207 commented on public/script.js:
```diff
@@ -2830,6 +2833,102 @@ export function scrollChatToBottom() {
+        return null;
+    }
+
+    const url = new URL(avatarSrc, window.location.origin);
```
Add a try-catch in case if the avatar src is not a valid URL

m-petra-fn replied:
Added.

m-petra-fn committed:
Re-rendering divs after other tag operations.

m-petra-fn committed:
Filtering message ids to only re-render specific divs. Adding suggested changes.

m-petra-fn committed:
After testing, it seems to add all messages by one, so it appears it's not needed call this in the printMessages function.

m-petra-fn committed:
Lint fixes.

Cohee1207 committed:
Cache jquery selector

Cohee1207 committed:
Fix local variables assignment

Cohee1207 committed:
Move new functions to tags module

Cohee1207 commented on public/scripts/tags.js:
```diff
@@ -2004,6 +2015,135 @@ function registerTagsSlashCommands() {
+function applyTags($element, tagData) {
+    $element.attr('data-char-tags', tagData.joinedTagNames);
+    tagData.tagNames.forEach(tagName => {
+        $element.attr(`data-char-tag-${tagName}`, '');
```
Here we would still need to sanitize attribute names, otherwise:

<img width="1098" height="169" alt="Image" src="https://github.com/user-attachments/assets/d4f77bfe-92c1-452c-9ce2-e32eaf45ad6e" />

m-petra-fn replied:
![2025-07-05 10_26_05-DevTools - 127 0 0 1_8000_](https://github.com/user-attachments/assets/7eb5a877-0910-449c-89b3-3f6dc83ba949)

CSS.escape didn't work as it would only make it `data-char-tag-ST\ Default`. So, I used regex to:

- Replace accented characters.
- Remove special chars except hyphens.
- Replace spaces and underscores with hyphens.
- Lowercased.

On data-char-tags the names are kept in their entirety, except for commas which are replaced with empty spaces since commas are the separator. Might be useful to select another separator instead of commas if those are unsuitable.

![image](https://github.com/user-attachments/assets/933de76d-4a9d-49c2-b45d-15627bf1e562)
![image](https://github.com/user-attachments/assets/54967347-9d7a-4374-8327-3376a7895efb)

Cohee1207 replied:
Commas are fine

m-petra-fn committed:
Normalizing tag names before adding them as data-*

m-petra-fn committed:
Replace commas with spaces to avoid issues with tag names containing commas.

m-petra-fn committed:
Correcting where the comma replacing should go.

Cohee1207 committed:
Add error handling for applyCharacterTagsToMessageDivs
